include/seqan3/alignment/pairwise/policy/affine_gap_policy.hpp

```diff
@@ -0,0 +1,101 @@
+
+namespace seqan3::detail
+{
+
+// ----------------------------------------------------------------------------
+// affine_gap_policy
+// ----------------------------------------------------------------------------
+
+/*!\brief The hot kernel implementation using affine gaps.
+ * \ingroup alignment_policy
+ * \tparam derived_t   The derived alignment algorithm.
+ * \tparam allocator_t The cell type of the dynamic programming matrix.
+ */
+template <typename derived_t, typename cell_t>
+class affine_gap_policy
+{
+protected:
+    /*!\name Member types
+     * \{
+     */
+    //!\brief The underlying score type.
+    using score_t = std::tuple_element_t<0, cell_t>;
+    //!\}
+
+    /*!\name Constructor, destructor and assignment
+     * \brief Defaulted all standard constructor.
+     * \{
+     */
+    constexpr affine_gap_policy()                                      noexcept = default;
+    constexpr affine_gap_policy(affine_gap_policy const &)             noexcept = default;
+    constexpr affine_gap_policy(affine_gap_policy &&)                  noexcept = default;
+    constexpr affine_gap_policy & operator=(affine_gap_policy const &) noexcept = default;
+    constexpr affine_gap_policy & operator=(affine_gap_policy &&)      noexcept = default;
+    ~affine_gap_policy()                                               noexcept = default;
+    //!}
+
+    /*!\brief Computes the score for one cell.
+     * \tparam        score_cell_type The type of the current cell.
+     * \tparam        chache_t        The type of the cache.
+     * \param[in,out] active_cell     The current cell in the dynamic programming matrix.
+     * \param[in,out] cache           The cache storing hot helper variables.
+     * \param[in]     score           The score of comparing the respective letters of the first and the second sequence.
+     */
+    template <typename score_cell_type, typename chache_t>
+    constexpr void compute_cell(score_cell_type & active_cell,
+                                chache_t & cache,
+                                score_t const score) const noexcept
+    {
+        // Unpack the cached variables.
+        auto & [main_score, hz_score] = active_cell;
+        auto & [prev_cell, gap_open, gap_extend] = cache;
+        auto & [prev_score, vt_score] = prev_cell;
+
+        // can we use GCC's builtin saturation arithmetics?
+        score_t tmp = prev_score + score;
+        tmp = std::max(tmp, vt_score);
+        tmp = std::max(tmp, hz_score);
+        prev_score = main_score;
+        main_score = tmp;
+        tmp += gap_open;
+        vt_score += gap_extend;
+        hz_score += gap_extend;
+        vt_score = std::max(vt_score, tmp);
+        hz_score = std::max(hz_score, tmp);
+    }
+
+    /*!\brief Creates the cache used for affine gap computation.
+     * \tparam    gap_scheme_t The type of the gap scheme.
+     * \param[in] scheme       The configured gap sheme.
+     * \returns The created cache.
```

Comment:
Please document the meaning of the tuple elements. For instance, you redefine gap_open here as open_score + gap_score, which is different from the conventions and should therefore be clearly described. 
I suggest to assign only gap_open here and in the above function (compute_cell) do the sum.

Reply:
Well, no the convention is actually that opening a gap is gap cost plus additional gap open costs. In SeqAn2 this was different. So I do the add in the cache generation in order to save the instruction in the hot kernel above.

Reply:
And yes as described before, I will make it a spearate class which is then documented and easier to follow.

Reply:
maybe also not call it `setup`? I would assume that any member function called `setup` somehow changes the object, but this doesn't.

Reply:
yeah, make_cache is better, although I still feel that it is something that should be local to the object and created automatically, but that's up to you.

Reply:
I'll revisit this when I do more benchmarks. This is all detail code anyway so I can refactor it later without breaking API